File: 1315_sum_of_nodes_with_even_valued_grandparent/Solution.cs
// LeetCode 1315 - Sum Of Nodes With Even Valued Grandparent
// https://leetcode.com/problems/sum-of-nodes-with-even-valued-grandparent/

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int val = 0, TreeNode left = null, TreeNode right = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class Solution {
    public int SumEvenGrandparent(TreeNode root) {
        int Dfs(TreeNode node, TreeNode parent, TreeNode grandparent) {
            if (node == null) return 0;
            int add = grandparent != null && grandparent.val % 2 == 0 ? node.val : 0;
            return add + Dfs(node.left, node, parent) + Dfs(node.right, node, parent);
        }
        return Dfs(root, null, null);
    }
}

File: 1315_sum_of_nodes_with_even_valued_grandparent/Solution.scala
// LeetCode 1315 - Sum of Nodes with Even-Valued Grandparent
// https://leetcode.com/problems/sum-of-nodes-with-even-valued-grandparent/

class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
  var value: Int = _value
  var left: TreeNode = _left
  var right: TreeNode = _right
}

object Solution {
  def sumEvenGrandparent(root: TreeNode): Int = {
    def dfs(node: TreeNode, parent: TreeNode, grandparent: TreeNode): Int = {
      if (node == null) return 0
      val add = if (grandparent != null && grandparent.value % 2 == 0) node.value else 0
      add + dfs(node.left, node, parent) + dfs(node.right, node, parent)
    }
    dfs(root, null, null)
  }
}
